### ISSUE TITLE
[datetime] Add support for skeletons

### ISSUE THREAD

**Issue by zbraniecki** [unicode-org/icu4x#272]
Initial landing of `DateTime` does not contain support for skeletons. We want them for ability to use `Components` and matching to patterns.

**Comment by gregtatum** [comment]
I'm interested in working on this after #107, although I see that @sffc is currently assigned to it. I would like to start with filling out this issue with links to prior art and discussions on the API design.

**Comment by sffc** [comment]
I changed you to assignee, @gregtatum :)

Here are some of my thoughts on skeletons:

### Predefined Common Skeletons

I would like to see fastpaths for the most common skeletons.  Here is a good starting point from the Google Closure i18n library:

https://github.com/google/closure-library/blob/3e6c510d016b9376bf8d34cb8179ef7aabc2d8b1/closure/goog/i18n/datetimeformat.js#L135

```
  FULL_DATE: 0,
  LONG_DATE: 1,
  MEDIUM_DATE: 2,
  SHORT_DATE: 3,
  FULL_TIME: 4,
  LONG_TIME: 5,
  MEDIUM_TIME: 6,
  SHORT_TIME: 7,
  FULL_DATETIME: 8,
  LONG_DATETIME: 9,
  MEDIUM_DATETIME: 10,
  SHORT_DATETIME: 11
```

I would also add month-day (date with no year/era), as I've seen that as a common feature request.

What do I mean by fastpath?  I think we can precompile these patterns as separate data provider keys.  For example, "dates/fulldatepattern@1" or maybe "datepatterns/fulldate@1" will return the specified pattern directly, without having to resolve skeletons at runtime.

I think this is important because:

1. We don't have to ship DTPG\* code for these use cases
2. Faster performance
3. Easier to add new ICU4X clients

\* *DTPG refers to DateTimePatternGenerator, the monolith of code in ICU (and soon ICU4X) that resolves datetime skeletons to datetime patterns*

### ECMA-402 Style Skeletons

ICU4C/J use strings to represent skeletons.  ECMA-402 uses option bags instead.

I think the ECMA-402 approach is superior, especially for Rust, in large part because we can do more logic to figure out what data we might need.  We can't resolve skeletons to patterns at compile time since they are locale-dependent, but we can at least tell what symbols\* the skeleton might require.  If the skeleton only requests date fields, then we don't need to include time symbols, for example.

We might want to supplement the ECMA-402 bag-based skeleton with a proc-macro that compiles the string-based skeleton to a bag.

\* *See #257 for an explanation of symbols data.*

### Spec Compliance

The UTS 35 spec for skeleton resolution is currently in a little bit of flux.  The biggest outstanding issue is [CLDR-13627](https://unicode-org.atlassian.net/browse/CLDR-13627), but there are others you can find on Jira.

### Do as much at compile time as possible

The ICU code for skeleton resolution (DTPG) is a bit of a mess, and it would be nice if we can clean it up.  Explore ways to do as much logic at build time (in CldrJsonDataProvider) as possible.  Make the data provider give you the most useful form of data possible, such that the code we ship in actual DateTimeFormat should be as simple and clean as possible.

**Comment by zbraniecki** [comment]
@sffc - can you help me understand what you mean by "fastpaths for common skeletons"?

Are you suggesting to have DataProvider fastpath for Option Bags that match certain skeletons?

Do you also suggest that we don't provide `options::Skeleton` side by side with `options::Bag` but instead provide a proc macro that takes `skeleton!()` and produces `options::Bag`?

**Comment by sffc** [comment]
> @sffc - can you help me understand what you mean by "fastpaths for common skeletons"?
> 
> Are you suggesting to have DataProvider fastpath for Option Bags that match certain skeletons?

That's one way of doing it.  What I more had in mind though would be a third method of instantiating a DateTimeFormat: datetime style (current), arbitrary skeleton (bag of fields), and predefined skeleton (one selection out of an enum with 10-15 choices).  However, we could still fastpath arbitrary skeletons into predefined skeletons if they match.

> Do you also suggest that we don't provide `options::Skeleton` side by side with `options::Bag` but instead provide a proc macro that takes `skeleton!()` and produces `options::Bag`?

That is what I am putting on the table for further discussion.

**Comment by gregtatum** [comment]
This comment is for my notes as I look into the prior art. I plan on editing it with my research.
s
### DateTimePatternGenerators
* [i18n/unicode/dtptngen.h
](https://github.com/unicode-org/icu/blob/master/icu4c/source/i18n/unicode/dtptngen.h)
* [i18n/dtptngen.cpp](https://github.com/unicode-org/icu/blob/master/icu4c/source/i18n/dtptngen.cpp)
* Example usage: [DateTimePatternGeneratorSample.java](https://github.com/unicode-org/icu/blob/master/icu4j/samples/src/com/ibm/icu/samples/text/datetimepatterngenerator/DateTimePatternGeneratorSample.java)

**Comment by gregtatum** [comment]
So here is an update on where I am with this:

I've done a lot of the research on the prior art, and understanding the terminology. I'm currently working on some local prototypes of different pieces of the architecture.

I've done a bit of prototyping around the serialization of the skeletons for use with the data providers. I'm not quite happy with what I have locally, so I'm going to explore the relationship between the components::Bag fields, skeleton representations, and pattern representations.

Today I'm going to start prototyping some of the skeleton matching algorithm, as I think the serialization could be driven by the needs of this algorithm. I don't really have concrete work that's worth showing yet, but I'm hopeful to comment back on some of the API design discussion.

I'd also like to make sure that #451 lands before writing any real code, but that's not blocking some early prototypes.

**Comment by gregtatum** [comment]
One thing I am discovering is that the components::Bag does not allow for every configuration of available [date field symbols](https://unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table).

I also wrote a script that collects every skeleton available in the CLDR, and how many patterns are available in the locale for it.

https://gist.github.com/gregtatum/1d76bbdb87132f71a969a10f0c1d2d9c#file-2-output-js

**Comment by gregtatum** [comment]
I think this issue needs better scoping, and a break out of separate issues. I added the discuss to add it to the meeting agenda. If we don't have time to discuss this week, I'll add my thoughts here.

**Comment by gregtatum** [comment]
The C-API for ICU4C provides a list of common skeletons. I think this was interesting enough to document for future work: https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/udat_8h.html

e.g.

```
#define | UDAT_YEAR   "y"
  | Constant for date skeleton with year.  More...
 
#define | UDAT_QUARTER   "QQQQ"
  | Constant for date skeleton with quarter.  More...
 
#define | UDAT_ABBR_QUARTER   "QQQ"
  | Constant for date skeleton with abbreviated quarter.  More...
 
#define | UDAT_YEAR_QUARTER   "yQQQQ"
  | Constant for date skeleton with year and quarter.  More...
 
#define | UDAT_YEAR_ABBR_QUARTER   "yQQQ"
  | Constant for date skeleton with year and abbreviated quarter.  More...
 
#define | UDAT_MONTH   "MMMM"
  | Constant for date skeleton with month.  More...

etc.
```
